### PR TITLE
ARROW-756: [C++] MSVC build fixes and cleanup, remove -fPIC flag from EP builds on Windows, Dev docs

### PR DESCRIPTION
**Pull request by wesm**
Includes existing patch for ARROW-757 and closes #492 

With this patch I'm able to build with

```
cmake -G "NMake Makefiles" -DCMAKE_BUILD_TYPE=Release ..
nmake
```

### CODE REVIEW

**Comment by wesm** [comment]
I did a clean build and am not able to reproduce the debug build linker issue, will report if it becomes reproducible

**Comment by wesm** [comment]
Appveyor build for latest: https://ci.appveyor.com/project/wesm/arrow/build/1.0.106

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -854,16 +864,8 @@ if(ARROW_IPC)
   endif()
 
   message(STATUS "Flatbuffers include dir: ${FLATBUFFERS_INCLUDE_DIR}")
-  message(STATUS "Flatbuffers static library: ${FLATBUFFERS_STATIC_LIB}")
   message(STATUS "Flatbuffers compiler: ${FLATBUFFERS_COMPILER}")
   include_directories(SYSTEM ${FLATBUFFERS_INCLUDE_DIR})
-  ADD_THIRDPARTY_LIB(flatbuffers
```
Comment (xhochy) [inline cpp/CMakeLists.txt]: Don't we need the library?

Comment (wesm) [inline cpp/CMakeLists.txt]: No, we header-only dependency, we never actually link to libflatbuffers

**Review by xhochy — APPROVED** [review]
Perfect!

**Comment by wesm** [comment]
Green build from Travis: https://travis-ci.org/wesm/arrow/builds/219383217. Thanks!

**Pull request closed without merging**